pennylane/templates/subroutines/all_singles_doubles.py

```diff
@@ -0,0 +1,202 @@
+    the third qubit. Similarly, the state :math:`\vert 0011 \rangle` corresponds to a double
+    excitation involving the four qubits.
+
+    This template initializes the quantum register of the :math:`n`-qubit system to
+    the Hartree-Fock state and applies :class:`~.pennylane.SingleExcitation` and
+    :class:`~.pennylane.DoubleExcitation` operations implemented as
+    `Givens rotations <https://en.wikipedia.org/wiki/Givens_rotation>`_ that act
+    in the subspace of two and four qubits, respectively.
+
+    The :class:`~.pennylane.SingleExcitation` operation performs a rotation
+    in the two-dimensional subspace :math:`\{\vert 10 \rangle, \vert 01 \rangle \}` of
+    the qubits ``[r, p]`` involved in the single excitation
+    :math:`\hat{c}_p^\dagger \hat{c}_r \vert \mathrm{HF} \rangle` of the initial state.
+    Similarly, the :class:`~.pennylane.DoubleExcitation` operations perform a rotation
+    in the subspace :math:`\{\vert 1100 \rangle, \vert 0011 \rangle \}` of
+    the qubits ``[s, r, q, p]`` that correspond to the double excitation
+    :math:`\hat{c}_p^\dagger \hat{c}_q^\dagger \hat{c}_r \hat{c}_s \vert \mathrm{HF} \rangle`.
+    The qubit indices ``s, r`` and ``q, p`` correspond, respectively, the occupied and unoccupied
+    spin-orbitals in the initial state, and :math:`c^\dagger` (:math:`c`) is the particle
+    creation (annihilation) operator. The total number of operations is determined by all
+    possible single- and double-excitations of the reference state.
+
+    For example, the quantum circuit for the case of two electrons and six qubits is sketched
+    in the figure below,
+
+    |
+
+    .. figure:: ../../_static/templates/subroutines/all_singles_doubles.png
+        :align: center
+        :width: 70%
+        :target: javascript:void(0);
+
+    |
+
+    In this case, we have four single- and double-excitations that preserves the total-spin
+    projection of the Hartree-Fock state. The :class:`~.pennylane.SingleExcitation` gates
+    :math:`G` act on the qubits ``[0, 2], [0, 4], [1, 3], [1, 5]`` as indicated by the
+    squares while the :class:`~.pennylane.DoubleExcitation` gates :math:`G^{(2)}` are applied
+    to the qubits ``[0, 1, 2, 3], [0, 1, 2, 5], [0, 1, 2, 4], [0, 1, 4, 5]``.
+
+    The resulting unitary conserves the number of particles and prepares the
+    :math:`n`-qubit system in a superposition of the initial HF state and
+    multiply-excited configurations.
+
+    Args:
+        weights (tensor_like): Size ``(len(singles) + len(doubles),)`` tensor containing the
+            angles :math:`\theta` entering the :class:`~.pennylane.SingleExcitation` and
+            :class:`~.pennylane.DoubleExcitation` operations. The indices of the qubits the
+            operations act on are generated with the :func:`~.excitations` function.
+        wires (Iterable): wires that the template acts on
+        hf_state (array[int]): Length ``len(wires)`` occupation-number vector representing the
```

Comment:
I think the template should just take as input the number of particles and number of qubits, and internally use the hf_state and singles, doubles functions

Reply:
This is a possibility. What I don't like of this approach is the following:

- For example, the user may want to define open-shell or spin-polarized reference states e.g., Ne=3, qubits=6 , |101010>. This is something that the ``hf_state`` function can not do at present . Having the qubits initialized inside the template will restrict its use to the RHF case.
- It is probably a good idea to keep the template's argument as free as possible from qchem-related inputs, like the number of electrons, spin-multiplicity, etc. I think that those things can be better implemented in the ``hf_state`` function.

Reply:
Could you not set `singles=None, doubles=None` and if it is not used, the template does the internal derivation, but if the user specifies these, they are used?

Reply:
It looks like this is too much automation which can be tricky. For example, depending on the reference state and the spin selection rule that users can specify in the ``excitations`` function, one can have only singles or doubles or both. I think it is convenient to keep these things separated.